libraries/tock-register-interface/src/expectations.rs

```diff
@@ -0,0 +1,324 @@
+                        val,
+                        n,
+                        *v
+                    ))
+                );
+                s.position += 1;
+            }
+            Some(Operation::Read(_, n, _)) => {
+                assert!(
+                    false,
+                    "{}",
+                    s.error(&std::format!(
+                        "Expected to read from {:?}, but wrote {:?} instead",
+                        n,
+                        inner.name
+                    ))
+                );
+            }
+            _ => {
+                assert!(
+                    false,
+                    "{}",
+                    s.error(&std::format!("No expectation at position {}", s.position))
+                );
+            }
+        }
+    }
+
+    /// Adds a modify expectation to the sequencer.
+    fn expect_modify(&self, field: FieldValue<Self::T, Self::R>) {
+        let inner = self.get_inner();
+        let reg = ptr::addr_of!(inner.name) as usize;
+        Sequencer::expect_modify(inner.sequencer, reg, inner.name, field.value.as_u64());
+    }
+}
+
+/// The `ExpectableInit` trait assists with initializing the macro-generated registers struct.
+pub trait ExpectableInit {
+    fn init(&mut self, sequencer: usize, name: &'static str);
+}
+
+impl<T: Expectable> ExpectableInit for T {
+    // Initialize a single `Expectable` struct.
+    fn init(&mut self, sequencer: usize, name: &'static str) {
+        Sequencer::inc_ref(sequencer);
+        let inner = self.get_inner_mut();
+        inner.sequencer = sequencer;
+        inner.name = name;
+    }
+}
+
+impl<T: Expectable, const N: usize> ExpectableInit for [T; N] {
+    // Initialize an array of `Expectable` structs.
+    fn init(&mut self, sequencer: usize, name: &'static str) {
+        for i in 0..N {
+            self[i].init(sequencer, name);
+        }
+    }
+}
+
+impl<const N: usize> ExpectableInit for [u8; N] {
+    // Initialize an array of u8.  This do-nothing implementation
+    // is needed to accomodate padding used by some register definitions.
+    fn init(&mut self, _sequencer: usize, _name: &'static str) {}
+}
+
+/// The `Operation` enum represents a register operation in the sequencer.
+/// Each operation encodes an opaque idenfier for the register, the name
+/// of the register and the value to be read or written.
+#[derive(Debug)]
+enum Operation {
+    Read(usize, &'static str, u64),
+    Modify(usize, &'static str, u64),
+    Write(usize, &'static str, u64),
+}
+
+/// The `Sequencer` holds a list of expected operations to be performed
+/// on a set of registers.
+#[derive(Default, Debug)]
+pub struct Sequencer {
+    sequence: Vec<Operation>,
+    position: usize,
+    refcount: isize,
+}
+
+// We keep a global list of sequencers so we don't have to track references,
+// mutabiltiy or lifetimes within the register structures.
+#[thread_local]
+static mut SEQUENCERS: Option<HashMap<usize, Sequencer>> = None;
```

Comment:
Use [`std::thread_local`](https://doc.rust-lang.org/std/macro.thread_local.html) to avoid needing the `thread_local` unstable feature.